File: src/matcher.ts
import { isPc, platformsCompatible } from "./platform.js";
import {
  CrossPlayPreference,
  DEFAULT_MATCHMAKING_CONFIG,
  MatchCandidate,
  MatchmakingConfig,
  MatchResult,
  Party,
  Platform,
  Player,
} from "./types.js";

export interface CompatibilityResult {
  compatible: boolean;
  reason?: string;
}

/** Check if a single candidate player can join a lobby defined by existing members. */
export function canJoinLobby(
  existingPlayers: Player[],
  candidate: Player,
): CompatibilityResult {
  if (existingPlayers.length === 0) {
    return { compatible: true };
  }

  for (const member of existingPlayers) {
    const memberAllowsCandidate = platformsCompatible(
      member.platform,
      member.crossPlayPreference,
      candidate.platform,
    );
    const candidateAllowsMember = platformsCompatible(
      candidate.platform,
      candidate.crossPlayPreference,
      member.platform,
    );

    if (!memberAllowsCandidate || !candidateAllowsMember) {
      const blockedSide = !memberAllowsCandidate ? member.id : candidate.id;
      return {
        compatible: false,
        reason: `Platform mismatch between ${blockedSide} and ${candidate.id} (PC excluded for console-only)`,
      };
    }
  }

  if (existingPlayers[0]?.region && candidate.region) {
    if (existingPlayers[0].region !== candidate.region) {
      return {
        compatible: false,
        reason: `Region mismatch: ${candidate.region} vs ${existingPlayers[0].region}`,
      };
    }
  }

  return { compatible: true };
}

/** Filter a pool down to players eligible for console-only Warzone lobbies (no PC). */
export function filterConsoleOnlyPool(players: Player[]): MatchResult {
  const accepted: MatchCandidate[] = [];
  const rejected: MatchResult["rejected"] = [];

  for (const player of players) {
    if (isPc(player.platform)) {
      rejected.push({
        playerId: player.id,
        reason: "PC players excluded from console-only matchmaking",
      });
      continue;
    }

    if (player.crossPlayPreference === CrossPlayPreference.SAME_PLATFORM) {
      // Still valid in pool, but tagged — matching happens at lobby formation.
      accepted.push({
        playerIds: [player.id],
        platforms: [player.platform],
        averageSkill: player.skillRating,
        region: player.region,
      });
      continue;
    }

    accepted.push({
      playerIds: [player.id],
      platforms: [player.platform],
      averageSkill: player.skillRating,
      region: player.region,
    });
  }

  return { accepted, rejected };
}

/** Resolve a party's effective platform rules — strictest preference wins. */
export function resolvePartyPreference(members: Player[]): CrossPlayPreference {
  const order = [
    CrossPlayPreference.SAME_PLATFORM,
    CrossPlayPreference.CONSOLE_ONLY,
    CrossPlayPreference.ALL,
  ];

  for (const pref of order) {
    if (members.some((m) => m.crossPlayPreference === pref)) {
      return pref;
    }
  }

  return CrossPlayPreference.CONSOLE_ONLY;
}

/** Validate an entire party can queue together under console-only rules. */
export function validateParty(party: Party, playersById: Map<string, Player>): CompatibilityResult {
  const members = party.memberIds
    .map((id) => playersById.get(id))
    .filter((p): p is Player => p !== undefined);

  if (members.length !== party.memberIds.length) {
    return { compatible: false, reason: "Party contains unknown player IDs" };
  }

  if (members.some((m) => isPc(m.platform))) {
    return {
      compatible: false,
      reason: "Party includes a PC player — cannot enter console-only queue",
    };
  }

  for (let i = 0; i < members.length; i++) {
    for (let j = i + 1; j < members.length; j++) {
      const a = members[i];
      const b = members[j];
      const partyPref = resolvePartyPreference(members);

      const aAllowsB = platformsCompatible(a.platform, partyPref, b.platform);
      const bAllowsA = platformsCompatible(b.platform, partyPref, a.platform);

      if (!aAllowsB || !bAllowsA) {
        return {
          compatible: false,
          reason: `Party members ${a.id} (${a.platform}) and ${b.id} (${b.platform}) cannot play together under ${partyPref}`,
        };
      }
    }
  }

  return { compatible: true };
}

/** Build squads from a player pool using console-only compatibility rules. */
export function formSquads(
  players: Player[],
  config: MatchmakingConfig = DEFAULT_MATCHMAKING_CONFIG,
): { squads: Player[][]; unmatched: Player[] } {
  const pool = [...players];
  const squads: Player[][] = [];
  const unmatched: Player[] = [];

  while (pool.length > 0) {
    const seed = pool.shift();
    if (!seed) break;

    const squad: Player[] = [seed];

    for (let i = pool.length - 1; i >= 0 && squad.length < config.squadSize; i--) {
      const candidate = pool[i];
      const { compatible } = canJoinLobby(squad, candidate);

      if (!compatible) continue;

      if (config.skillBand > 0 && seed.skillRating !== undefined && candidate.skillRating !== undefined) {
        if (Math.abs(seed.skillRating - candidate.skillRating) > config.skillBand) {
          continue;
        }
      }

      squad.push(candidate);
      pool.splice(i, 1);
    }

    if (squad.length === config.squadSize) {
      squads.push(squad);
    } else {
      unmatched.push(...squad);
    }
  }

  return { squads, unmatched };
}

/** Form a full lobby from squads / solo players. */
export function formLobby(
  players: Player[],
  config: MatchmakingConfig = DEFAULT_MATCHMAKING_CONFIG,
): { lobby: Player[]; overflow: Player[] } {
  const { accepted, rejected } = filterConsoleOnlyPool(players);

  if (rejected.length > 0) {
    // PC already filtered — continue with accepted solo entries mapped back to Player objects.
  }

  const eligibleIds = new Set(accepted.flatMap((c) => c.playerIds));
  const eligible = players.filter((p) => eligibleIds.has(p.id));

  const lobby: Player[] = [];
  const overflow: Player[] = [];

  for (const player of eligible) {
    if (lobby.length >= config.lobbySize) {
      overflow.push(player);
      continue;
    }

    const { compatible } = canJoinLobby(lobby, player);
    if (compatible) {
      lobby.push(player);
    } else {
      overflow.push(player);
    }
  }

  return { lobby, overflow };
}

/** Summarize platform mix in a lobby — useful for telemetry / UI. */
export function lobbyPlatformSummary(players: Player[]): Record<Platform, number> {
  const summary: Record<Platform, number> = {
    [Platform.PC]: 0,
    [Platform.XBOX_SERIES]: 0,
    [Platform.XBOX_ONE]: 0,
    [Platform.PS5]: 0,
    [Platform.PS4]: 0,
  };

  for (const p of players) {
    summary[p.platform]++;
  }

  return summary;
}

/** True when lobby has zero PC players and at least one console platform. */
export function isConsoleOnlyLobby(players: Player[]): boolean {
  return players.length > 0 && players.every((p) => !isPc(p.platform));
}

/** Preset config for Warzone console players who want Xbox + PS but no PC. */
export function consoleOnlyWarzoneConfig(
  overrides: Partial<MatchmakingConfig> = {},
): MatchmakingConfig {
  return {
    ...DEFAULT_MATCHMAKING_CONFIG,
    ...overrides,
  };
}

export function createConsoleOnlyPlayer(
  id: string,
  platform: Exclude<Platform, Platform.PC>,
  options: Partial<Omit<Player, "id" | "platform">> = {},
): Player {
  return {
    id,
    platform,
    crossPlayPreference: CrossPlayPreference.CONSOLE_ONLY,
    ...options,
  };
}

File: src/matcher.cpp
#include "warzone/matcher.hpp"

#include "warzone/platform.hpp"

#include <algorithm>
#include <cmath>
#include <set>
#include <sstream>

namespace warzone {

namespace {

MatchCandidate makeCandidate(const Player& player) {
  MatchCandidate candidate;
  candidate.playerIds = {player.id};
  candidate.platforms = {player.platform};
  candidate.averageSkill = player.skillRating;
  candidate.region = player.region;
  return candidate;
}

}  // namespace

CompatibilityResult canJoinLobby(const std::vector<Player>& existingPlayers,
                                 const Player& candidate) {
  if (existingPlayers.empty()) {
    return {true, std::nullopt};
  }

  for (const Player& member : existingPlayers) {
    const bool memberAllowsCandidate = platformsCompatible(
        member.platform, member.crossPlayPreference, candidate.platform);
    const bool candidateAllowsMember = platformsCompatible(
        candidate.platform, candidate.crossPlayPreference, member.platform);

    if (!memberAllowsCandidate || !candidateAllowsMember) {
      const std::string blockedSide =
          !memberAllowsCandidate ? member.id : candidate.id;
      std::ostringstream reason;
      reason << "Platform mismatch between " << blockedSide << " and "
             << candidate.id << " (PC excluded for console-only)";
      return {false, reason.str()};
    }
  }

  if (existingPlayers.front().region && candidate.region &&
      *existingPlayers.front().region != *candidate.region) {
    std::ostringstream reason;
    reason << "Region mismatch: " << *candidate.region << " vs "
           << *existingPlayers.front().region;
    return {false, reason.str()};
  }

  return {true, std::nullopt};
}

MatchResult filterConsoleOnlyPool(const std::vector<Player>& players) {
  MatchResult result;

  for (const Player& player : players) {
    if (isPc(player.platform)) {
      result.rejected.push_back(
          {player.id, "PC players excluded from console-only matchmaking"});
      continue;
    }

    result.accepted.push_back(makeCandidate(player));
  }

  return result;
}

CrossPlayPreference resolvePartyPreference(const std::vector<Player>& members) {
  const CrossPlayPreference order[] = {CrossPlayPreference::SamePlatform,
                                       CrossPlayPreference::ConsoleOnly,
                                       CrossPlayPreference::All};

  for (CrossPlayPreference pref : order) {
    if (std::any_of(members.begin(), members.end(),
                    [pref](const Player& member) {
                      return member.crossPlayPreference == pref;
                    })) {
      return pref;
    }
  }

  return CrossPlayPreference::ConsoleOnly;
}

CompatibilityResult validateParty(
    const Party& party,
    const std::unordered_map<std::string, Player>& playersById) {
  std::vector<Player> members;
  members.reserve(party.memberIds.size());

  for (const std::string& memberId : party.memberIds) {
    const auto it = playersById.find(memberId);
    if (it == playersById.end()) {
      return {false, "Party contains unknown player IDs"};
    }
    members.push_back(it->second);
  }

  if (std::any_of(members.begin(), members.end(),
                  [](const Player& member) { return isPc(member.platform); })) {
    return {false, "Party includes a PC player — cannot enter console-only queue"};
  }

  const CrossPlayPreference partyPref = resolvePartyPreference(members);

  for (size_t i = 0; i < members.size(); ++i) {
    for (size_t j = i + 1; j < members.size(); ++j) {
      const Player& a = members[i];
      const Player& b = members[j];

      const bool aAllowsB =
          platformsCompatible(a.platform, partyPref, b.platform);
      const bool bAllowsA =
          platformsCompatible(b.platform, partyPref, a.platform);

      if (!aAllowsB || !bAllowsA) {
        std::ostringstream reason;
        reason << "Party members " << a.id << " (" << platformToString(a.platform)
               << ") and " << b.id << " (" << platformToString(b.platform)
               << ") cannot play together under "
               << crossPlayPreferenceToString(partyPref);
        return {false, reason.str()};
      }
    }
  }

  return {true, std::nullopt};
}

SquadFormationResult formSquads(const std::vector<Player>& players,
                                const MatchmakingConfig& config) {
  SquadFormationResult result;
  std::vector<Player> pool = players;

  while (!pool.empty()) {
    Player seed = pool.front();
    pool.erase(pool.begin());

    std::vector<Player> squad = {seed};

    for (int i = static_cast<int>(pool.size()) - 1;
         i >= 0 && static_cast<int>(squad.size()) < config.squadSize; --i) {
      const Player& candidate = pool[static_cast<size_t>(i)];
      const CompatibilityResult compatibility = canJoinLobby(squad, candidate);
      if (!compatibility.compatible) {
        continue;
      }

      if (config.skillBand > 0 && seed.skillRating && candidate.skillRating) {
        if (std::abs(*seed.skillRating - *candidate.skillRating) > config.skillBand) {
          continue;
        }
      }

      squad.push_back(candidate);
      pool.erase(pool.begin() + i);
    }

    if (static_cast<int>(squad.size()) == config.squadSize) {
      result.squads.push_back(std::move(squad));
    } else {
      result.unmatched.insert(result.unmatched.end(), squad.begin(), squad.end());
    }
  }

  return result;
}

LobbyFormationResult formLobby(const std::vector<Player>& players,
                               const MatchmakingConfig& config) {
  LobbyFormationResult result;
  const MatchResult filtered = filterConsoleOnlyPool(players);

  std::set<std::string> eligibleIds;
  for (const MatchCandidate& candidate : filtered.accepted) {
    for (const std::string& playerId : candidate.playerIds) {
      eligibleIds.insert(playerId);
    }
  }

  for (const Player& player : players) {
    if (eligibleIds.find(player.id) == eligibleIds.end()) {
      continue;
    }

    if (static_cast<int>(result.lobby.size()) >= config.lobbySize) {
      result.overflow.push_back(player);
      continue;
    }

    const CompatibilityResult compatibility = canJoinLobby(result.lobby, player);
    if (compatibility.compatible) {
      result.lobby.push_back(player);
    } else {
      result.overflow.push_back(player);
    }
  }

  return result;
}

std::unordered_map<Platform, int> lobbyPlatformSummary(
    const std::vector<Player>& players) {
  std::unordered_map<Platform, int> summary = {
      {Platform::Pc, 0},
      {Platform::XboxSeries, 0},
      {Platform::XboxOne, 0},
      {Platform::Ps5, 0},
      {Platform::Ps4, 0},
  };

  for (const Player& player : players) {
    ++summary[player.platform];
  }

  return summary;
}

bool isConsoleOnlyLobby(const std::vector<Player>& players) {
  return !players.empty() &&
         std::all_of(players.begin(), players.end(),
                     [](const Player& player) { return !isPc(player.platform); });
}

MatchmakingConfig consoleOnlyWarzoneConfig(const MatchmakingConfig& overrides) {
  MatchmakingConfig config = defaultMatchmakingConfig();
  config.squadSize = overrides.squadSize;
  config.lobbySize = overrides.lobbySize;
  config.requireSameRegion = overrides.requireSameRegion;
  config.skillBand = overrides.skillBand;
  return config;
}

Player createConsoleOnlyPlayer(const std::string& id,
                               Platform platform,
                               const Player& defaults) {
  Player player = defaults;
  player.id = id;
  player.platform = platform;
  player.crossPlayPreference = CrossPlayPreference::ConsoleOnly;
  return player;
}

}  // namespace warzone
